tests_func/load/mongo_load.go

```diff
@@ -0,0 +1,275 @@
+	var res MongoRunCommandOp
+	bmrco, err := json.Marshal(cd)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse opts for commad: %+v", err)
+	}
+	err = json.Unmarshal(bmrco, &res)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse opts for commad: %+v", err)
+	}
+	err = bson.UnmarshalExtJSON(cd["dc"], true, &res.Doc)
+	if err != nil {
+		return nil, err
+	}
+	return &res, nil
+}
+
+func mongoRunCommandOp(ctx context.Context, client *mongo.Client, op *MongoRunCommandOp) string {
+	db := client.Database(op.DbName)
+	var result bson.M
+	err := db.RunCommand(ctx, op.Doc).Decode(&result)
+	if err != nil {
+		return fmt.Sprintf("Failed execution of runCommand with id %d: %+v", op.Id, err)
+	} else {
+		return fmt.Sprintf("Successful execution of runCommand with id %d", op.Id)
+	}
+}
+
+func addFuncCommand(client *mongo.Client, op *MongoRunCommandOp) func(ctx context.Context) string {
+	return func(ctx context.Context) string {
+		return mongoRunCommandOp(ctx, client, op)
+	}
+}
+
+type MongoRunTransactionOp struct {
+	Cmds     []map[string]json.RawMessage `json:"cmds"`
+	Id       int                          `json:"id"`
+	callback func(ctx context.Context, client *mongo.Client) (interface{}, error)
+}
+
+func parseTransactionOp(cd map[string]json.RawMessage) (*MongoRunTransactionOp, error) {
+	var res MongoRunTransactionOp
+	bmrco, err := json.Marshal(cd)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse opts for commad: %+v", err)
+	}
+	err = json.Unmarshal(bmrco, &res)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse opts for commad: %+v", err)
+	}
+	var fns []func(ctx context.Context, client *mongo.Client) string
+	for _, cmd := range res.Cmds {
+		commandDoc, err := parseCommandOp(cmd)
+		if err != nil {
+			return nil, err
+		}
+		fns = append(fns, func(ctx context.Context, client *mongo.Client) string {
+			return mongoRunCommandOp(ctx, client, commandDoc)
+		})
+	}
+	res.callback = func(ctx context.Context, client *mongo.Client) (interface{}, error) {
+		var res string
+		for _, f := range fns {
+			res = res + "\n\t|\t" + f(ctx, client)
+		}
+		return res, nil
+	}
+	return &res, nil
+}
+
+func mongoFuncTransaction(client *mongo.Client, op *MongoRunTransactionOp) func(ctx context.Context) string {
+	return func(ctx context.Context) string {
+		session, err := client.StartSession()
+		if err != nil {
+			return fmt.Sprintf("Cannot start session: %+v", err)
+		}
+		defer session.EndSession(context.Background())
+		tres, err := session.WithTransaction(ctx, func(sessCtx mongo.SessionContext) (interface{}, error) {
+			return op.callback(ctx, client)
+		})
+		if err != nil {
+			return fmt.Sprintf("Failed execution of transaction with id %d: %+v", op.Id, err)
+		} else {
+			return fmt.Sprintf("Successful execution of transaction with id %d%s", op.Id, tres)
+		}
+	}
+}
+
+func MongoRunFuncs(ctx context.Context, fc <-chan func(ctx context.Context) string, resSize int) (<-chan string, <-chan struct{}) {
```

Comment:
Is it possible to rewrite as mentioned in "Bounded parallelism" https://blog.golang.org/pipelines ?

Reply:
Done